File: data structures and algorithms/java/08_013.java
import java.util.HashMap;
import java.util.Map;

public class FirstNonRepeatingChar {
    public static int firstUniqChar(String s) {
        Map<Character, Integer> charCount = new HashMap<>();
        
        // Count the occurrences of each character
        for (char c : s.toCharArray()) {
            charCount.put(c, charCount.getOrDefault(c, 0) + 1);
        }
        
        // Find the first character with count 1
        for (int i = 0; i < s.length(); i++) {
            if (charCount.get(s.charAt(i)) == 1) {
                return i;
            }
        }
        
        return -1;  // No unique character found
    }
    
    public static void main(String[] args) {
        System.out.println(firstUniqChar("apple"));      // 0
        System.out.println(firstUniqChar("loveapples")); // 1
        System.out.println(firstUniqChar("aabb"));       // -1
    }
}

File: data structures and algorithms/cpp/08_013.cpp
#include <iostream>
#include <string>
#include <unordered_map>

int firstUniqChar(std::string s) {
    std::unordered_map<char, int> charCount;
    
    // Count the occurrences of each character
    for (char c : s) {
        charCount[c]++;
    }
    
    // Find the first character with count 1
    for (int i = 0; i < s.length(); i++) {
        if (charCount[s[i]] == 1) {
            return i;
        }
    }
    
    return -1;  // No unique character found
}

int main() {
    std::cout << firstUniqChar("apples") << std::endl;      // 0
    std::cout << firstUniqChar("loveapples") << std::endl;  // 1
    std::cout << firstUniqChar("aabb") << std::endl;        // -1
    return 0;
}